src/core/unit.c

```diff
@@ -5396,27 +5401,35 @@ int unit_prepare_exec(Unit *u) {
 
 static void log_leftover(pid_t pid, int sig, void *userdata) {
         _cleanup_free_ char *comm = NULL;
+        UnitResult *ur = (UnitResult *) userdata;
 
         (void) get_process_comm(pid, &comm);
 
         if (comm && comm[0] == '(') /* Most likely our own helper process (PAM?), ignore */
                 return;
 
-        log_unit_warning(userdata,
+        log_unit_warning(ur->u,
                          "Found left-over process " PID_FMT " (%s) in control group while starting unit. Ignoring.\n"
                          "This usually indicates unclean termination of a previous run, or service implementation deficiencies.",
                          pid, strna(comm));
+
+        ur->result = true;
+
+        return;
 }
 
-void unit_warn_leftover_processes(Unit *u) {
+bool unit_warn_leftover_processes(Unit *u) {
         assert(u);
+        UnitResult ur = { .u = u, .result = false };
 
         (void) unit_pick_cgroup_path(u);
 
         if (!u->cgroup_path)
-                return;
+                return false;
+
+        (void) cg_kill_recursive(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, 0, 0, NULL, log_leftover, &ur);
 
-        (void) cg_kill_recursive(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, 0, 0, NULL, log_leftover, u);
+        return ur.result;
```

Comment:
Hmm, I think it would be nicer if the logging callback that cg_kill_recursive() takes would be slightly altered, so that it returns a boolean. And then cg_kill_recursive() should return > 0 when at least one invocation of that function returns true, and 0 otherwise. Right now cg_kill_recursive() already returns > 0 when at least one process is killed, and letting the log function decide whether to return that would be only a minor extension the way I see it, and a natural one.

With that in place you don't need the small structure to pass data between this function and the log function, which would be nicer I think.